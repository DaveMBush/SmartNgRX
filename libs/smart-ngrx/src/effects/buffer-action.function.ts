import { NgZone } from '@angular/core';
import { Action } from '@ngrx/store';
import {
  asapScheduler,
  buffer,
  debounceTime,
  map,
  Observable,
  Subscriber,
} from 'rxjs';

import { castTo } from '../common/cast-to.function';

function mainBuffer(
  source: Observable<Action>,
  bufferTime: number,
  ngZone: NgZone,
  observer: Subscriber<string[]>,
) {
  source
    .pipe(
      map((a) => castTo<{ ids: string[] }>(a).ids),
      buffer(source.pipe(debounceTime(bufferTime, asapScheduler))),
      map((ids: string[][]) => {
        let newIds: string[] = [];
        ids.forEach((ids2) => {
          newIds = [...newIds, ...ids2];
        });
        return newIds;
      }),
      map((ids) => ids.filter((c, index) => ids.indexOf(c) === index)),
    )
    .subscribe({
      next: (value) => ngZone.run(() => observer.next(value)),
      error: (err: unknown) => ngZone.run(() => observer.error(err)),
      complete: () => ngZone.run(() => observer.complete()),
    });
}

/**
 * This is an internal function that is used by the Effects to buffer
 * IDs of an action coming into an effect so that we can dispatch
 * them independently but send them to the server in a single request.
 *
 * NOTE: bufferAction assumes an array of ids is passed to the action
 * it is buffering.
 *
 * ## Usage:
 *
 * ``` typescript
 *  load$ = createEffect(
      (
        actions$ = inject(Actions),
        actionService = inject(effectServiceToken),
        zone: NgZone = inject(NgZone)
      ) => {
        return actions$.pipe(
          ofType(actions.loadByIds),
          bufferAction(zone), // <--- buffer the ids
          mergeMap((ids) => actionService.loadByIds(ids)),
          map((rows) => actions.loadByIdsSuccess({ rows }))
        );
      },
      { functional: true }
    );
 * ```
 * @param ngZone - The zone to use to run outside of Angular.
 * @param bufferTime - The time to buffer the ids before sending them to the server.
 *     The default is 1ms which only allow the buffer to last until the thread frees up
 *     and is probably all we will ever need.
 * @returns The buffered ids.
 */
export function bufferAction(
  ngZone: NgZone,
  bufferTime = 1,
): (source: Observable<Action>) => Observable<string[]> {
  return (source: Observable<Action>): Observable<string[]> => {
    return new Observable<string[]>((observer) => {
      ngZone.runOutsideAngular(() =>
        mainBuffer(source, bufferTime, ngZone, observer),
      );
    });
  };
}
