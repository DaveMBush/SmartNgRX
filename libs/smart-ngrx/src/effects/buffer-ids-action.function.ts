import { NgZone } from '@angular/core';
import { Action } from '@ngrx/store';
import {
  asapScheduler,
  buffer,
  debounceTime,
  map,
  Observable,
  Subscriber,
} from 'rxjs';

import { forNext } from '../common/for-next.function';

function flatten<T>(array: T[][]): T[] {
  const returnArray = [] as T[];
  forNext(array, (a) => {
    forNext(a, (b) => {
      returnArray.push(b);
    });
  });
  return returnArray;
}

function mainIdsBuffer(
  source: Observable<Action & { ids: string[] }>,
  bufferTime: number,
  ngZone: NgZone,
  observer: Subscriber<string[]>,
) {
  source
    .pipe(
      map((a) => a.ids),
      buffer(source.pipe(debounceTime(bufferTime, asapScheduler))),
      map((ids: string[][]) => {
        return flatten(ids);
      }),
      map((ids) => ids.filter((c, index) => ids.indexOf(c) === index)),
    )
    .subscribe({
      next: (value) => ngZone.run(() => observer.next(value)),
      error: (err: unknown) => ngZone.run(() => observer.error(err)),
      complete: () => ngZone.run(() => observer.complete()),
    });
}

/**
 * This is an internal function that is used by the Effects to buffer
 * IDs of an action coming into an effect so that we can dispatch
 * them independently but send them to the server in a single request.
 *
 * NOTE: bufferAction assumes an array of ids is passed to the action
 * it is buffering.
 *
 * ## Usage:
 *
 * ``` typescript
 * load$ = createEffect(
 * (
 * actions$ = inject(Actions),
 * actionService = inject(effectServiceToken),
 * zone: NgZone = inject(NgZone)
 * ) => {
 * return actions$.pipe(
 * ofType(actions.loadByIds),
 * bufferAction(zone), // <--- buffer the ids
 * mergeMap((ids) => actionService.loadByIds(ids)),
 * map((rows) => actions.loadByIdsSuccess({ rows }))
 * );
 * },
 * { functional: true }
 * );
 * ```
 *
 * @param ngZone The zone to use to run outside of Angular.
 * @param bufferTime The time to buffer the ids before sending them to the server.
 *     The default is 1ms which only allow the buffer to last until the thread frees up
 *     and is probably all we will ever need.
 * @returns The buffered ids.
 */
export function bufferIdsAction(
  ngZone: NgZone,
  /* istanbul ignore next */
  bufferTime = 1, // default value does not need to be tested
): (source: Observable<Action & { ids: string[] }>) => Observable<string[]> {
  return (
    source: Observable<Action & { ids: string[] }>,
  ): Observable<string[]> => {
    return new Observable<string[]>((observer) => {
      ngZone.runOutsideAngular(() =>
        mainIdsBuffer(source, bufferTime, ngZone, observer),
      );
    });
  };
}
