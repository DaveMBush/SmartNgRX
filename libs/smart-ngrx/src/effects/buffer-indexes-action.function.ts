import { NgZone } from '@angular/core';
import { Action } from '@ngrx/store';
import {
  asapScheduler,
  buffer,
  debounceTime,
  groupBy,
  map,
  mergeMap,
  Observable,
  Subscriber,
} from 'rxjs';

import { forNext } from '../common/for-next.function';
import { IndexesProp } from '../types/indexes-props.interface';

function flatten(prop: IndexesProp[]): IndexesProp {
  const returnProps = {} as IndexesProp;
  forNext(prop, (a) => {
    forNext(a.indexes, (b) => {
      returnProps.indexes.push(b);
    });
  });
  return returnProps;
}

function mainIndexesBuffer(
  source: Observable<Action & IndexesProp>,
  bufferTime: number,
  ngZone: NgZone,
  observer: Subscriber<IndexesProp>,
) {
  source
    .pipe(
      groupBy(action => `${action.parentId}-${action.childField}`),
      mergeMap(grouped =>
        grouped.pipe(
          buffer(grouped.pipe(debounceTime(bufferTime, asapScheduler))),
          map((actions: IndexesProp[]) => flatten(actions))
        )
      )
    )
    .subscribe({
      next: (value) => ngZone.run(() => observer.next(value)),
      error: (err: unknown) => ngZone.run(() => observer.error(err)),
      complete: () => ngZone.run(() => observer.complete()),
    });
}
/**
 * This is an internal function that is used by the Effects to buffer
 * IDs of an action coming into an effect so that we can dispatch
 * them independently but send them to the server in a single request.
 *
 * NOTE: bufferAction assumes an array of ids is passed to the action
 * it is buffering.
 *
 * ## Usage:
 *
 * ``` typescript
 * load$ = createEffect(
 * (
 * actions$ = inject(Actions),
 * actionService = inject(effectServiceToken),
 * zone: NgZone = inject(NgZone)
 * ) => {
 * return actions$.pipe(
 * ofType(actions.loadByIds),
 * bufferAction(zone), // <--- buffer the ids
 * mergeMap((ids) => actionService.loadByIds(ids)),
 * map((rows) => actions.loadByIdsSuccess({ rows }))
 * );
 * },
 * { functional: true }
 * );
 * ```
 *
 * @param ngZone The zone to use to run outside of Angular.
 * @param bufferTime The time to buffer the ids before sending them to the server.
 *     The default is 1ms which only allow the buffer to last until the thread frees up
 *     and is probably all we will ever need.
 * @returns The buffered ids.
 */
export function bufferIndexesAction(
  ngZone: NgZone,
  /* istanbul ignore next */
  bufferTime = 1, // default value does not need to be tested
): (source: Observable<Action & IndexesProp>) => Observable<IndexesProp> {
  return (
    source: Observable<Action & IndexesProp>,
  ): Observable<IndexesProp> => {
    return new Observable<IndexesProp>((observer) => {
      ngZone.runOutsideAngular(() =>
        mainIndexesBuffer(source, bufferTime, ngZone, observer),
      );
    });
  };
}
