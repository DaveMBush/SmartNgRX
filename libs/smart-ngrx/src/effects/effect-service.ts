import { Observable } from 'rxjs';

/**
 * This is the abstract class that all services the Effects
 * use must implement.
 */
export abstract class EffectService<T> {
  /**
   * Used to load ALL the rows from the server for the given entity.
   */
  abstract load: () => Observable<T[]>;
  /**
   * Loads the rows represented by the array of ids passed in.
   */
  abstract loadByIds: (ids: string[]) => Observable<T[]>;

  abstract update: (oldRow: T, newRow: T) => Observable<T[]>;
}
