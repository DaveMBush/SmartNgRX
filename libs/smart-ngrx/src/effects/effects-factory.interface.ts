import { LoadByIdsEffect } from './load-by-ids-effect.type';
import { LoadEffect } from './load-effect.type';
/**
 * This is an internal interface that should not be required
 * by anyone developing with this lib.
 *
 * This is the type information that effectsFactory returns.
 *
 * @see `effectsFactory`
 * @see `EffectService`
 * @see `LoadEffect`
 * @see `LoadByIdsEffect`
 */
export interface EffectsFactory<T> {
  /**
   * The effect the `load` action will end up calling.
   */
  load: LoadEffect<T>;
  /**
   * Used internally to make the store think the records
   * represented by the IDs has already been loaded by using
   * the defaultRow functionality.
   */
  loadByIdsPreload: LoadEffect<T>;
  /**
   * The effect the `loadById` action will end up calling.
   */
  loadByIds: LoadByIdsEffect<T>;
}
