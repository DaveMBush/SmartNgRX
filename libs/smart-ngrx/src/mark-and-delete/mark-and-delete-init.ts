import { assert } from '../common/assert.function';
import { isNullOrUndefined } from '../common/is-null-or-undefined.function';
import { MarkAndDeleteInit } from '../types/mark-and-delete-init.interface';

const markAndDeleteRecord = {} as Record<string, MarkAndDeleteInit>;

/**
 * Function that allows us to register the `markAndDeleteInit` globally
 * and for each entity.
 *
 * @param feature the feature to register the mark and delete init for
 * @param markAndDelete the mark and delete init for the entity
 */
export function registerMarkAndDeleteInit(
  feature: string,
  markAndDelete: MarkAndDeleteInit,
): void {
  markAndDeleteRecord[feature] = markAndDelete;
}

/**
 * retrieves previously registered `MarkAndDeleteInit` for the entity
 *
 * @param entity the entity or ${feature}:${entity} to retrieve
 * @returns the `MarkAndDeleteInit` for the entity
 */
export function getMarkAndDeleteInit(entity: string): MarkAndDeleteInit {
  assert(
    !isNullOrUndefined(markAndDeleteRecord[entity]),
    `MarkAndDeleteInit for ${entity} is not registered`,
  );
  return markAndDeleteRecord[entity];
}
