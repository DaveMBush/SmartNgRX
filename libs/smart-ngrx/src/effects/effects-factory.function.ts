import { InjectionToken } from '@angular/core';
import { createEffect, EffectConfig, FunctionalEffect } from '@ngrx/effects';
import { EntityAdapter } from '@ngrx/entity';

import { actionFactory } from '../actions/action.factory';
import { assert } from '../common/assert.function';
import { castTo } from '../common/cast-to.function';
import { StringLiteralSource } from '../ngrx-internals/string-literal-source.type';
import { entityDefinitionCache } from '../registrations/entity-definition-cache.function';
import { SmartNgRXRowBase } from '../types/smart-ngrx-row-base.interface';
import { EffectService } from './effect-service';
import { addEffect } from './effects-factory/add-effect.function';
import { addSuccessEffect } from './effects-factory/add-success-effect.function';
import { deleteEffect } from './effects-factory/delete-effect.function';
import { loadByIdsEffect } from './effects-factory/load-by-ids-effect.function';
import { loadByIdsPreloadEffect } from './effects-factory/load-by-ids-preload-effect.function';
import { loadEffect } from './effects-factory/load-effect.function';
import { updateEffect } from './effects-factory/update-effect.function';

const dispatchFalse = {
  dispatch: false,
  functional: true,
} as EffectConfig & {
  functional: true;
  dispatch: false;
};

const dispatchTrue = {
  functional: true,
} as EffectConfig & {
  functional: true;
  dispatch: true;
};

/**
 * The effects factory creates a new set of effects for the
 * `Action` source provided and calls the service represented
 * by the `InjectionToken` provided.
 *
 * @param feature the feature name this effect is being run for
 * @param entityName the entity within the feature this effect is being run for
 * @param effectsServiceToken The token for the service that
 *   the resulting effect will call.
 * @returns The effect for the source provided
 *
 * @see `EffectsFactory`
 * @see `EffectService`
 */
export function effectsFactory<
  F extends string,
  E extends string,
  T extends SmartNgRXRowBase,
>(
  feature: StringLiteralSource<F>,
  entityName: StringLiteralSource<E>,
  effectsServiceToken: InjectionToken<EffectService<T>>,
): Record<string, FunctionalEffect> {
  const actions = actionFactory<T, F, E>(feature, entityName);
  const adapter = castTo<EntityAdapter<T> | undefined>(
    entityDefinitionCache(feature, entityName).entityAdapter,
  );
  assert(!!adapter, `Missing adapter for ${feature}:${entityName}.`);
  return castTo<Record<string, FunctionalEffect>>({
    delete: createEffect(
      deleteEffect(effectsServiceToken, actions),
      dispatchFalse,
    ),
    load: createEffect(loadEffect(effectsServiceToken, actions), dispatchTrue),
    loadByIdsPreload: createEffect(
      loadByIdsPreloadEffect(feature, entityName, actions),
      dispatchFalse,
    ),
    loadByIds: createEffect(
      loadByIdsEffect(
        effectsServiceToken,
        actions,
        feature as StringLiteralSource<string>,
        entityName as StringLiteralSource<string>,
      ),
      dispatchFalse,
    ),
    update: createEffect(
      updateEffect<T>(
        effectsServiceToken,
        actions,
        feature as StringLiteralSource<string>,
        entityName as StringLiteralSource<string>,
      ),
      dispatchFalse,
    ),
    add: createEffect(addEffect(effectsServiceToken, actions), dispatchTrue),
    addSuccess: createEffect(
      addSuccessEffect(effectsServiceToken, actions, adapter),
      dispatchFalse,
    ),
  });
}
