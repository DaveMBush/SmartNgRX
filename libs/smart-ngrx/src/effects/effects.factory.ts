import { InjectionToken } from '@angular/core';
import { createEffect, FunctionalEffect } from '@ngrx/effects';

import { castTo } from '../common/cast-to.function';
import { actionFactory } from '../functions/action.factory';
import { StringLiteralSource } from '../ngrx-internals/string-literal-source.type';
import { MarkAndDelete } from '../types/mark-and-delete.interface';
import { EffectService } from './effect-service';
import { loadByIdsEffect } from './effects-factory/load-by-ids-effect.function';
import { loadByIdsPreloadEffect } from './effects-factory/load-by-ids-preload-effect.function';
import { loadEffect } from './effects-factory/load-effect.function';

/**
 * The effects factory creates a new set of effects for the
 * `Action` source provided and calls the service represented
 * by the `InjectionToken` provided.
 *
 * @param source The source of the actions for this effect
 * @param feature
 * @param entityName
 * @param effectsServiceToken The token for the service that
 *   the resulting effect will call.
 * @returns The effect for the source provided
 *
 * @see `EffectsFactory`
 * @see `EffectService`
 */
export function effectsFactory<
  F extends string,
  E extends string,
  T extends MarkAndDelete,
>(
  feature: StringLiteralSource<F>,
  entityName: StringLiteralSource<E>,
  effectsServiceToken: InjectionToken<EffectService<T>>,
): Record<string, FunctionalEffect> {
  const actions = actionFactory<F, E, T>(feature, entityName);
  return castTo<Record<string, FunctionalEffect>>({
    load: createEffect(loadEffect(effectsServiceToken, actions), {
      functional: true,
    }),
    loadByIdsPreload: createEffect(loadByIdsPreloadEffect(actions), {
      functional: true,
    }),
    loadByIds: createEffect(loadByIdsEffect(effectsServiceToken, actions), {
      functional: true,
    }),
  });
}
