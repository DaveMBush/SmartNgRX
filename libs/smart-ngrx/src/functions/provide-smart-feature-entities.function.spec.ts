import { Injectable, InjectionToken } from '@angular/core';
import { Observable, of } from 'rxjs';

import { global } from '../common/global.const';
import { EffectService } from '../effects/effect-service';
import { getMarkAndDeleteInit } from '../mark-and-delete/mark-and-delete-init';
import { StringLiteralSource } from '../ngrx-internals/string-literal-source.type';
import { MarkAndDelete } from '../types/mark-and-delete.interface';
import { SmartEntityDefinition } from '../types/smart-entity-definition.interface';
import { provideSmartFeatureEntities } from './provide-smart-feature-entities.function';
import { provideSmartNgRX } from './provide-smart-ngrx.function';
import { unregisterEntity } from './register-entity.function';

interface Test extends MarkAndDelete {
  id: string;
  isDirty: boolean;
}

@Injectable()
class TestEffectsService extends EffectService<Test> {
  override load: () => Observable<Test[]> = () => {
    return of([] as Test[]);
  };

  override loadByIds: (ids: string[]) => Observable<Test[]> = (
    ids: string[],
  ) => {
    return of(ids.map((id) => ({ id, isDirty: false }) as Test));
  };
}

describe('provideSmartFeatureEntities', () => {
  let result: unknown;
  let featureName = '';
  let entityDefinition: SmartEntityDefinition<Test>;
  afterEach(() => {
    unregisterEntity(featureName, entityDefinition.fieldName);
  });

  describe('when we define a feature/entity with no mark and delete config', () => {
    beforeEach(() => {
      featureName = 'someFeature';
      entityDefinition = {
        fieldName: 'reducerName',
        effectServiceToken: new InjectionToken<TestEffectsService>(
          'TestEffectsService',
        ),
        defaultRow: (id: string) => ({ id, isDirty: false }),
      };
      const entityDefinitions: SmartEntityDefinition<MarkAndDelete>[] = [
        entityDefinition,
      ];
      // sets the default mark and delete config
      provideSmartNgRX();
      result = provideSmartFeatureEntities(
        featureName as StringLiteralSource<typeof featureName>,
        entityDefinitions,
      );
    });
    it('should provide the NgRX reducer and effect for this slice and use the global mark and delete information', () => {
      expect(result).toMatchSnapshot();
      expect(
        getMarkAndDeleteInit(`${featureName}:${entityDefinition.fieldName}`),
      ).toBe(getMarkAndDeleteInit(global));
    });
  });
  describe('when we define a feature/entity with a mark and delete config', () => {
    beforeEach(() => {
      featureName = 'someFeature';
      entityDefinition = {
        fieldName: 'reducerName',
        effectServiceToken: new InjectionToken<TestEffectsService>(
          'TestEffectsService',
        ),
        defaultRow: (id: string) => ({ id, isDirty: false }),
        markAndDelete: {
          markDirtyTime: 3000,
          removeTime: 2000,
          runInterval: 1000,
          markDirtyFetchesNew: false,
        },
      };
      const entityDefinitions: SmartEntityDefinition<MarkAndDelete>[] = [
        entityDefinition,
      ];
      // sets the default mark and delete config
      provideSmartNgRX();
      result = provideSmartFeatureEntities(
        featureName as StringLiteralSource<typeof featureName>,
        entityDefinitions,
      );
    });
    it("should provide the NgRX reducer and effect for this slice and use the entity's mark and delete information", () => {
      expect(result).toMatchSnapshot();
      expect(
        getMarkAndDeleteInit(`${featureName}:${entityDefinition.fieldName}`),
      ).toStrictEqual({
        markDirtyTime: 3000,
        removeTime: 2000,
        runInterval: 1000,
        markDirtyFetchesNew: false,
      });
    });
  });
});
