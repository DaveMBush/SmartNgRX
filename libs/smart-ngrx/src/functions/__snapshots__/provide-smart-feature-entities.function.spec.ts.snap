// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provideSmartFeatureEntities when we define a feature/entity with a mark and delete config should provide the NgRX reducer and effect for this slice and use the entity's mark and delete information 1`] = `
{
  "ɵfromNgModule": true,
  "ɵproviders": [
    {
      "deps": [],
      "provide": [Function],
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "INJECTOR_DEF_TYPES",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "ENVIRONMENT_INITIALIZER",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "deps": [],
      "provide": [Function],
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "INJECTOR_DEF_TYPES",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "ENVIRONMENT_INITIALIZER",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Configs",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {},
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Store Features",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {
        "initialState": undefined,
        "key": "someFeature",
        "metaReducers": [],
        "reducerFactory": [Function],
      },
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Feature Configs",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        InjectionToken {
          "_desc": "@ngrx/store Store Features",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Store Features",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {
        "reducerName": [Function],
      },
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers Token",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useExisting": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Feature Reducers",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        [
          ParamDecoratorFactory {
            "token": InjectionToken {
              "_desc": "@ngrx/store Internal Feature Reducers Token",
              "ngMetadataName": "InjectionToken",
              "ɵprov": undefined,
            },
          },
        ],
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Runtime Checks",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Check if Action types are unique",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects Internal Feature Effects",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [
        {
          "load": [Function],
          "loadByIds": [Function],
          "loadByIdsPreload": [Function],
        },
      ],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects User Provided Effects",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [],
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/effects Internal Feature Effects",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        InjectionToken {
          "_desc": "@ngrx/effects User Provided Effects",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects Internal Feature Effects Instance Groups",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
  ],
}
`;

exports[`provideSmartFeatureEntities when we define a feature/entity with no mark and delete config should provide the NgRX reducer and effect for this slice and use the global mark and delete information 1`] = `
{
  "ɵfromNgModule": true,
  "ɵproviders": [
    {
      "deps": [],
      "provide": [Function],
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "INJECTOR_DEF_TYPES",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "ENVIRONMENT_INITIALIZER",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "deps": [],
      "provide": [Function],
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "INJECTOR_DEF_TYPES",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "ENVIRONMENT_INITIALIZER",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Configs",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {},
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Store Features",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {
        "initialState": undefined,
        "key": "someFeature",
        "metaReducers": [],
        "reducerFactory": [Function],
      },
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Feature Configs",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        InjectionToken {
          "_desc": "@ngrx/store Store Features",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Store Features",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": {
        "reducerName": [Function],
      },
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers Token",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useExisting": InjectionToken {
        "_desc": "@ngrx/store Internal Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Feature Reducers",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        [
          ParamDecoratorFactory {
            "token": InjectionToken {
              "_desc": "@ngrx/store Internal Feature Reducers Token",
              "ngMetadataName": "InjectionToken",
              "ɵprov": undefined,
            },
          },
        ],
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Feature Reducers",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/store Internal Runtime Checks",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/store Check if Action types are unique",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects Internal Feature Effects",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [
        {
          "load": [Function],
          "loadByIds": [Function],
          "loadByIdsPreload": [Function],
        },
      ],
    },
    {
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects User Provided Effects",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useValue": [],
    },
    {
      "deps": [
        InjectionToken {
          "_desc": "@ngrx/effects Internal Feature Effects",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
        InjectionToken {
          "_desc": "@ngrx/effects User Provided Effects",
          "ngMetadataName": "InjectionToken",
          "ɵprov": undefined,
        },
      ],
      "multi": true,
      "provide": InjectionToken {
        "_desc": "@ngrx/effects Internal Feature Effects Instance Groups",
        "ngMetadataName": "InjectionToken",
        "ɵprov": undefined,
      },
      "useFactory": [Function],
    },
  ],
}
`;
